packages/hardhat-core/src/internal/cli/autocomplete.ts

```diff
@@ -97,7 +102,58 @@ export async function complete({
     .map(ArgumentsParser.paramNameToCLA)
     .filter((x) => !words.includes(x));
 
-  return [...taskParams, ...coreParams];
+  return [...taskParams, ...coreParams].filter((completion) =>
+    completion.startsWith(last)
+  );
+}
+
+function filesystemSuggestions(last: string): string[] {
+  let partialDirectory: string;
+  let partialFilename: string;
+
+  const parsedPath = path.parse(last);
+
+  // something like foo/. is technically a directory, but we don't want to consider it as such
+  if (parsedPath.base !== "." && isDirectory(last) && last.endsWith(path.sep)) {
+    partialDirectory = last;
+    partialFilename = "";
+  } else {
+    partialDirectory = parsedPath.dir;
+    partialFilename = parsedPath.base;
+  }
+
+  const directory = path.resolve(process.cwd(), partialDirectory);
+
+  const suggestions = fs
+    .readdirSync(directory)
+    .filter((filename) => filename.startsWith(partialFilename))
+    .filter(
+      (filename) => partialFilename.startsWith(".") || !filename.startsWith(".")
+    )
+    .map((filename) => path.join(partialDirectory, filename));
+
+  // dirty trick when we match a single directory:
+  // we don't want to have a single suggestion because otherwise the shell
+  // will add an space at the end, which is annoying,
+  // so we suggest both the dir, and the dir with a trailing path separator
+  if (suggestions.length === 1) {
+    const [suggestion] = suggestions;
+
+    if (isDirectory(suggestion)) {
+      return [suggestion, path.join(suggestion, path.sep)];
```

Comment:
I'm not proud of this, but it works surprisingly well.

Reply:
lol